src/librustc_trans/back/write.rs

```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
-        let remark = sess.opts.cg.remark.clone();
-
-        let (tx, rx) = channel();
-        let mut tx = Some(tx);
-        futures.push(rx);
-
-        let incr_comp_session_dir = sess.incr_comp_session_dir_opt().map(|r| r.clone());
-
-        let depth = time_depth();
-        thread::Builder::new().name(format!("codegen-{}", i)).spawn(move || {
-            set_time_depth(depth);
-
-            let diag_handler = Handler::with_emitter(true, false, box diag_emitter);
-
-            // Must construct cgcx inside the proc because it has non-Send
-            // fields.
-            let cgcx = CodegenContext {
-                lto_ctxt: None,
-                handler: &diag_handler,
-                plugin_passes: plugin_passes,
-                remark: remark,
-                worker: i,
-                incr_comp_session_dir: incr_comp_session_dir
-            };
+fn execute_work<'a>(sess: &'a Session,
+                    mut work_items: Vec<WorkItem>,
+                    jobserver: Client,
+                    exported_symbols: &'a ExportedSymbols,
+                    scope: &Scope<'a>) {
+    let (tx, rx) = channel();
+    let tx2 = tx.clone();
+
+    // First up, convert our jobserver into a helper thread so we can use normal
+    // mpsc channels to manage our messages and such. Once we've got the helper
+    // thread then request `n-1` tokens because all of our work items are ready
+    // to go.
+    //
+    // Note that the `n-1` is here because we ourselves have a token (our
+    // process) and we'll use that token to execute at least one unit of work.
+    let helper = jobserver.into_helper_thread(move |token| {
+        drop(tx2.send(Message::Token(token)));
+    }).expect("failed to spawn helper thread");
+    for _ in 0..work_items.len() - 1 {
+        helper.request_token();
+    }
 
-            loop {
-                // Avoid holding the lock for the entire duration of the match.
-                let maybe_work = work_items_arc.lock().unwrap().pop();
-                match maybe_work {
-                    Some(work) => {
-                        execute_work_item(&cgcx, work);
+    let mut tokens = Vec::new();
+    let mut running = 0;
+    while work_items.len() > 0 || running > 0 {
+        // On each turn of this loop start of by trying to spawn as much work as
+        // possible. Remember that we have an ambient token available to us
+        // hence the `+1` here.
+        //
+        // Also note that we may actually acquire more tokens than we need, so
+        // in that case just truncate the `tokens` list every time we pass
+        // through here.
+        while work_items.len() > 0 && running < tokens.len() + 1 {
+            spawn_work(sess,
+                       exported_symbols,
+                       scope,
+                       tx.clone(),
+                       work_items.pop().unwrap(),
+                       work_items.len());
```

Comment:
I'm not very fond this: mutating `work_items` via `pop` and then taking its `len`. I assume that we have a defined evaluation order of function arguments, but I don't like relying on it.